File: 2080-range-frequency-queries/2080-range-frequency-queries.java
class RangeFreqQuery {
    
    class SegmentTree {
        Map<Integer, Integer>[] seg;
        int[] arr;
        SegmentTree(int[] arrcpy) {
            arr = new int[arrcpy.length];
            System.arraycopy(arrcpy, 0, arr, 0, arr.length);
            seg = new Map[4 * arr.length];
            for(int i=0; i<seg.length; i++) {
                seg[i] = new HashMap<>();
            }
            buildTree(0, 0, arrcpy.length-1);
        }
        void buildTree(int idx, int left, int right) {
            if(left == right) {
                seg[idx].put(arr[left], seg[idx].getOrDefault(arr[left], 0) + 1);
                return;
            }
            int mid = left + (right - left) / 2;
            buildTree(2 * idx + 1, left, mid);
            buildTree(2 * idx + 2, mid + 1, right);
            for(Map.Entry<Integer, Integer> entry : seg[2 * idx + 1].entrySet()) {
                seg[idx].put(entry.getKey(), seg[idx].getOrDefault(entry.getKey(), 0) + entry.getValue());
            }
            for(Map.Entry<Integer, Integer> entry : seg[2 * idx + 2].entrySet()) {
                seg[idx].put(entry.getKey(), seg[idx].getOrDefault(entry.getKey(), 0) + entry.getValue());
            }
        }
        int query(int idx, int left, int right, int l, int r, int val) {
            if(l > right || r < left) return 0;
            if(l <= left && right <= r) return seg[idx].getOrDefault(val, 0);
            int mid = left + (right - left) / 2;
            return query(2 * idx + 1, left, mid, l, r, val) + query(2 * idx + 2, mid + 1, right, l, r, val);
        }
    }

    int len;
    SegmentTree obj;
    public RangeFreqQuery(int[] arr) {
        len = arr.length-1;
        obj = new SegmentTree(arr);
        //System.out.println(Arrays.toString(obj.seg));
    }
    
    public int query(int left, int right, int value) {
        return obj.query(0, 0, len, left, right, value);
    }
}

/**
 * Your RangeFreqQuery object will be instantiated and called as such:
 * RangeFreqQuery obj = new RangeFreqQuery(arr);
 * int param_1 = obj.query(left,right,value);
 */

File: 2080-range-frequency-queries/2080-range-frequency-queries.py
class RangeFreqQuery:
    arr1 = []
    m = dict()
    def __init__(self, arr: List[int]):
        self.arr1 = arr
        self.m = dict()
        for i in range(len(arr)):
            if arr[i] not in self.m:
                self.m[arr[i]] = []
            self.m[arr[i]].append(i)

    def query(self, left: int, right: int, value: int) -> int:
        if value not in self.m: return 0
        one = self.bSearch(0, len(self.m[value])-1, self.m[value], left)
        two = self.bSearch(0, len(self.m[value])-1, self.m[value], right)
        if one < 0 and two < 0 and one == two: return 0
        one = -one-1 if one < 0 else one
        two = -two-2 if two < 0 else two
        return two - one + 1

    def bSearch(self, left: int, right: int, arr: List[int], val: int) -> int:
        while left <= right:
            mid = left + (right - left) // 2
            if arr[mid] == val: return mid
            elif val < arr[mid]: right = mid - 1
            else: left = mid + 1
        return -left-1
    
# Your RangeFreqQuery object will be instantiated and called as such:
# obj = RangeFreqQuery(arr)
# param_1 = obj.query(left,right,value)